### ISSUE TITLE
FileStream throws exception

### ISSUE THREAD

**Issue by godrose** [cureos/shim#51]
Using the latest preview version 2.1.0-pre throws an exception when trying to open a file:
System.PlatformNotSupportedException : PCL

   at System.IO.FileStream..ctor(String path, FileMode mode) in d:\Users\Anders\Documents\Visual Studio 2013\Projects\shim\Sources\Shim\IO\FileStream.cs:line 32

```
using (var fileStream = new FileStream(SerializedBuildersPath, FileMode.Create))
{
}
```

The project which contains the line is portable with the Profile111

**Comment by anders9ustafsson** [comment]
The application is on which platform? I think your application is missing an _explicit_ reference to _Shim_. Please double-check that your application is including the _Shim_ NuGet package explicitly.

**Comment by anders9ustafsson** [comment]
@godrose Did you manage to resolve this?

**Comment by godrose** [comment]
Hi @anders9ustafsson 
Sorry for the late response. Just checked on this one -still reproduces.
The setup is the following:
I have a PCL which references the 2.0.0 Shim version with portable profile. This library is contained inside the correspondent nuget package. 
There is a solution which uses this nuget package, also referencing the Shim explicitly.
The project which uses the package is also portable with the same profile.
The Shim is taken from portable-net45+netcore45+wpa81 folder.

**Comment by anders9ustafsson** [comment]
Hi @godrose !

But I assume that you encounter the error when you run an _application_ that references your PCL, and the application is running on a specific platform, e.g. .NET or Xamarin iOS?

It is **not** enough that your PCL references _Shim_. For the _bait-and-switch trick_ to work, you **also** need to explicitly reference _Shim_ from your application, preferably by adding the _Shim_ NuGet package directly to your application. Only then will the target specific _Shim_ (with proper _implementations_ of the `FileStream` class and other types) be selected by your application.

**Comment by godrose** [comment]
It is added as a reference.
I will expand on the setup:
There's a Library A which references the 2.0.0 Shim version with portable profile. This library is contained inside the correspondent nuget package. 
The solution containing this Library A also contains a test project whose target framework is 4.5.1.
These tests run just fine.

There is another solution which contains a portable project B that has the Library A added as a nuget package. There's a test project C which references the portable project B and has Shim installed as a nuget package. The test project C has 4.5.1 as the target framework. In this case the exception is thrown.

If the project B is reverted to target 4.5.1 everything works fine.

_Update_:  This does not reproduce anymore.

**Comment by anders9ustafsson** [comment]
@godrose OK, thanks!

If you look at the properties of the _Shim_ library in the test project C's references, which path does this library have? To a _portable_ or _net_ folder?

Have you re-targeted project C at any time? Please remove and re-add the _Shim_ package to project C and check whether this affects the project C _Shim_ path.

**Comment by godrose** [comment]
1) it's the **net** folder.
2) No, it hasn't been retargeted. It has the correct path.

**Comment by anders9ustafsson** [comment]
OK, strange? And the same code works when you run unit tests on library A, you say?

**Comment by godrose** [comment]
Yes. I will test more scenario and tell you if I find something interesting.

**Comment by godrose** [comment]
Hasn't been able to reproduce consistently.
One interesting finding is that re-building the solution occasionally results in the 'Not supported exception'. Cleaning the output folder and re-building again solves the issue.

**Comment by tlhintoq** [comment]
Same error for me.
Profile 111 (Xamarin Forms solution)
Projects for: PCL, Android, iOS, UWP
PCL has Shim 2.0.1 (no pre-release showing in NuGet manager for the PCL project)
UWP has Shim 2.1.0-pre

When I try to clone an object through BinaryFormatter and MemoryStream, the same PlatformNotSupported error is thrown.

```
        public static T Clone<T>(T source)
        {
            // Not available in PCL
            //if (!typeof(T).IsSerializable)
            //{
            //    throw new ArgumentException("The type must be serializable.", "source");
            //}

            // Don't serialize a null object, simply return the default for that object
            if (Object.ReferenceEquals(source, null))
            {
                return default(T);
            }

            var formatter = new BinaryFormatter();
            Stream stream = new MemoryStream();
            using (stream)
            {
                formatter.Serialize(stream, source);
                stream.Seek(0, SeekOrigin.Begin);
                return (T)formatter.Deserialize(stream);
            }
        }

```

**Comment by tlhintoq** [comment]
@anders9ustafsson 

Just wondering if any progress has been made on this.  I'm finding it still happens with 2.1.10-pre.

In an odd manifestation of it it... BinaryFormatter.Strea.Serialize throws the same `PlatformNotSupported`.  When you view the details of the exception it claims Shim is not referenced: Yet it is in any way I can find to see it referenced.  See attached.  If I try to explicitly add a reference to the dll I can't because it is already automatically referenced.

[![2017-01-19_10-25-56.png](https://s6.postimg.org/6nhnc6341/2017_01_19_10_25_56.png)](https://postimg.org/image/58g2ng20t/)

**Comment by anders9ustafsson** [comment]
@tlhintoq Seems like it is trying to reference the PCL Profile 328 version when it really should reference the UWP version? Strange? 

What happens if you remove all Shim references and try to re-add them?

**Comment by tlhintoq** [comment]
@anders9ustafsson After the last issue ([#57](https://github.com/cureos/shim/issues/57)) that was the first thing I tried.  NuGet manager for solution.  Pulled Shim out of the entire solution (all projects) then reinstalled it making sure they all got the same -pre version.
